BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,395 @@
+
+            if (!std::filesystem::exists(texure_path))
+            {
+                THROW_EX("textrue image doesn't exist on specified path")
+            }
+
+            Texture::Ptr tex = image_io->LoadImage(light.texture);
+            ibl->SetTexture(tex);
+            ibl->SetMultiplier(light.mul);
+        }
+        else
+        {
+            THROW_EX("unsupported light type")
+        }
+
+        light_instance->SetPosition(light.pos);
+        light_instance->SetDirection(light.dir);
+        light_instance->SetEmittedRadiance(light.rad);
+        m_scene->AttachLight(light_instance);
+    }
+}
+
+void Render::GenerateDataset(const std::vector<CameraInfo>& camera_states,
+                             const std::vector<LightInfo>& light_settings,
+                             const std::vector<int>& spp,
+                             const std::filesystem::path& output_dir,
+                             bool gamma_corection_enabled)
+{
+    using namespace RadeonRays;
+
+    if (!std::filesystem::is_directory(output_dir))
+    {
+        THROW_EX("incorrect output directory signature");
+    }
+
+    // check if number of samples to render wasn't specified
+    if (spp.empty())
+    {
+        return;
+    }
+
+    const int iter_number = *(std::max_element(spp.begin(), spp.end()));
+
+    SetLight(light_settings);
+
+    int cam_index = 1;
+    for (const auto &cam_state: camera_states)
+    {
+        // create camera if it wasn't  done earlier
+        if (!m_camera)
+        {
+            m_camera = Baikal::PerspectiveCamera::Create(cam_state.at,
+                                                         cam_state.pos,
+                                                         cam_state.up);
+
+            m_camera->SetSensorSize(RadeonRays::float2(0.036f, 0.036f));
+            m_camera->SetDepthRange(RadeonRays::float2(0.0f, 100000.f));
+
+            m_scene->SetCamera(m_camera);
+        }
+
+        UpdateCameraSettings(cam_state);
+
+        for (const auto& output: m_outputs)
+        {
+            output->Clear(RadeonRays::float3());
+        }
+
+        // recompile scene cause of changing camera pos and settings
+        m_controller->CompileScene(m_scene);
+        auto& scene = m_controller->GetCachedScene(m_scene);
+
+        for (auto i = 1; i <= iter_number; i++)
+        {
+            m_renderer->Render(scene);
+
+            if (i == 1)
+            {
+                for (const auto& output : kSingleIteratedOutputs)
+                {
+                    std::stringstream ss;
+
+                    ss << "cam_" << cam_index << "_"
+                        << output.name << ".bin";
+
+                    SaveOutput(output,
+                               ss.str(),
+                               gamma_corection_enabled,
+                               output_dir);
+                }
+            }
+
+            if (std::find(spp.begin(), spp.end(), i) != spp.end())
```

Comment:
This check can be costly and can be optimized since at every new loop iteration the actual number of elements we need to inspect in spp container is diminishing, so it would make sense to collect spps into linear memory array and then advance staring tip every time we encounter target spp.

Reply:
Done